File: Java/Problem1-BestTimeToBuyAndSellStock3.java
class Solution {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int m = prices[0];
        int b = 0;
        int dp = new int[n];

        dp[0] = 0;
        for (int i = 1; i < n; i++) {
            if (b < prices[i] - m) b = prices[i] - m;
            if (m > prices[i]) m = prices[i];

            dp[i] = b;
        }

        int M = prices[n - 1];
        int B = 0;
        int ans = dp[n - 1];

        for (int i  = n - 2; i > 0; i--) {
            if (B < M - prices[i]) B = M - prices[i];
            if (M < prices[i]) M = prices[i];

            if (ans < dp[i - 1] + B) ans = B + dp[i - 1];
        }

        return ans;
    }
}

File: Python/Problem1-BestTimeToBuyAndSellStock3.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        n = len(prices)
        m = prices[0]
        b = 0
        dp = [0]

        for price in prices:
            if b < price - m:
                b = price - m
            if m > price:
                m = price

            dp.append(b)

        M = prices[n - 1]
        B = 0
        ans = dp[n]

        for i in range(n - 1, 0, -1):
            price = prices[i]
            if B < M - price:
                B = M - price
            if M < price:
                M = price
            
            if ans < dp[i] + B:
                ans = dp[i] + B
        
        return ans
